### PR TITLE
Edit/Create - Documentation refactor

### PR DESCRIPTION
**Pull request by riccardozambito**
Library documentation refactor:
- Updated readme.md
- Created a folder [./docs](./docs) where are stored the MD files of each doc section ( a macro topic about the library) .
-  Each section contains links to the Platform docs and specific examples.
- With this PR the name of the library becomes **Mia Node.js service library**

#### Checklist

- [x] your branch will not cause merge conflict with `master`
- [x] run `npm test`
- [x] tests are included
- [x] the documentation is updated or integrated accordingly with your changes
- [x] the code will follow the lint rules and style of the project
- [x] you are not committing extraneous files or sensible data

### CODE REVIEW

**Comment by giulioroggero** [comment]
The supported node.js version is 12 lts

File: docs/HTTPClient.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * `delete(path, body, querystring, options)`
+
+The params to be passed to these functions are:
+
+ * `path` -  a string that identifies the route to which you want to send the request.
+ * `body` - optional, the body of the request which can be:
+    * a JSON object
+    * a [Buffer](https://nodejs.org/api/buffer.html#)
+    * one [Stream](https://nodejs.org/api/stream.html)
+ * `querystring` - optional, an object that represents the querystring.
```
Comment (fredmaggiowski) [inline docs/HTTPClient.md:37]:
```suggestion
 * `querystring` - optional, an object that represents the querystring;
```

File: docs/Logging.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# Logging
+You can log a message to see in DevOps console. The library use the [Fastify logging system](https://www.fastify.io/docs/v2.0.x/Logging/), that is based on [pino](https://github.com/pinojs/pino).
+
+To log messages call these methods on the logger instance of request. Logging is enabled by default. Therefore you can call on `request.log`:
```
Comment (fredmaggiowski) [inline docs/Logging.md:4]:
```suggestion
To log messages call these methods on the logger instance of request. Logging is enabled by default. Therefore you can call on `request.log` or `service.log`:
```

File: docs/Logging.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+* `debug()`
+* `info()`
+* `warn()`
+* `error()`
+* `fatal()`
+
+Each method creates a log with the homonym level.
+
```
Comment (fredmaggiowski) [inline docs/Logging.md:12]: By default the library will generate two logs for each request, one representing the incoming request and one for the request completion, logs are created with trace and info levels respectively and already provide useful information for later analysis or debugging, if you need more you can add your own logs.

File: docs/Logging.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+```js
+service.addPostDecorator('/notify', function notifyHandler(request) {
+  // Get "notifications" setting from the request querystring
+  const { notifications } = request.getOriginalRequestQuery()
+  //
+  if(!notifications)  {
+    return req.leaveOriginalResponseUnmodified()
+  }
+  else {
```
Comment (fredmaggiowski) [inline docs/Logging.md:23]: This else is useless

File: docs/Logging.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+Each method creates a log with the homonym level.
+
+## Example
+
+```js
+service.addPostDecorator('/notify', function notifyHandler(request) {
+  // Get "notifications" setting from the request querystring
+  const { notifications } = request.getOriginalRequestQuery()
+  //
```
Comment (fredmaggiowski) [inline docs/Logging.md:19]:
```suggestion

```

File: docs/Logging.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  if(!notifications)  {
+    return req.leaveOriginalResponseUnmodified()
+  }
+  else {
+      try {
+      const notifier=new Notifier()
```
Comment (fredmaggiowski) [inline docs/Logging.md:25]:
```suggestion
      const notifier = new Notifier()
```

File: docs/Logging.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      const response = await notifier.send({ text: `${who} says: ${mymsg}`})
+      const sendedAt = new Date();
+
+      // Log at "INFO" level
+      req.log.info({ statusCode: response.statusCode }, 'Notify sent')
+
+      return request.changeOriginalResponse().setBody(
+        {...req.getOriginalRequestBody(),notifySendedAt:sendedAt})
```
Comment (fredmaggiowski) [inline docs/Logging.md:33]:
```suggestion
        { ...req.getOriginalRequestBody(), notifySendedAt:sendedAt}
    )
```

File: README.md [hunk 8]
```diff
@@ -10,30 +10,91 @@
...
-Is highly copled with [`lc39`][lc39] and on [Fastify][fastify].
+Built on [`Fastify`][fastify], it takes advantage of Mia-Platform Node.js service launcher [`lc39`][lc39].
+
+# Getting Started
+You can use this module in your projects or, from the [DevOps Console](https://docs.mia-platform.eu/development_suite/overview-dev-suite/), get started quickly and easily with a [ready-to-use microservice template](https://docs.mia-platform.eu/development_suite/api-console/api-design/custom_microservice_get_started/). Even in  the [Mia-Platform Marketplace](https://github.com/mia-platform-marketplace) repository, you can find some examples and boilerplates that using this library.
+
+## Setup the local development environment
+
+To develop the service locally you need:
+- Node.js v12 LTS or later.
```
[the remark this answers is not in the document]
Comment (riccardozambito) [reply README.md:23]: Yes, i think it's fair.

File: docs/HTTPClient.md (hunk 1 above)
Comment (riccardozambito) [reply docs/HTTPClient.md:37]: I don't understand why using ";" in this case. Can you explain to me?

File: docs/CustomService.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+const customService = require('@mia-platform/custom-plugin-lib')({
+  type: 'object',
+  required: ['ENV_VAR'],
+  properties: {
+    ENV_VAR: { type: 'string' },
+  },
+})
+```
+
+> **_More examples?_** Go [here](../examples/advanced/index.js#L46) to see another use cases.
```
[the remark this answers is not in the document]
Comment (riccardozambito) [reply docs/CustomService.md:19]: I thought to quote it to better highlight it.

File: docs/Decorators.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+service.addPostDecorator('/notify', function notifyHandler(request) {
+  // Get "notifications" setting from the request querystring
+  const { notifications } = request.getOriginalRequestQuery()
+    //
+    if(!notifications)  {
+      // It's not necessary to send notification
+      // leave the original response unmodified
+      return req.leaveOriginalResponseUnmodified()
+    }
+    else {
```
[the remark this answers is not in the document]
Comment (riccardozambito) [reply docs/Decorators.md]: Sorry, an old habit.

File: README.md [hunk 11]
```diff
@@ -65,8 +65,8 @@ module.exports = customService(async function helloWorldService(service) {
   })
 }) 
 ```
-- The library exports a function, `customService`, that expects an async function to initialize and configure the `service`. Optionally can take a schema of the required environment variables (you can find the reference [here][fastify-env]).
-- `service` is a [Fastify instance](https://www.fastify.io/docs/latest/Server/), upon which, you can call `addRawCustomPlugin` method that allows you to add your route. You can use  *service* to register any Fastify plugin, see [here][fastify-ecosystem] for a list of currently available plugins.
+- The library exports a function that returns another functions. Optionally can take a schema of the required environment variables (you can find the reference [here][fastify-env]). The function returned, `customService`, expects an async function to initialize and configure the `service`.
```
Comment (fredmaggiowski) [inline README.md:68]: Maybe we should clarify better this sentence... Is one of the first interesting things a user reads about the plugin and I think it's a bit too technical without giving useful information, for instance: `The library exports a function that returns another functions.` it's true but still a user reading this line won't learn much on how to use the library..

Maybe there's a better way to express this concept?

File: README.md [hunk 12]
```diff
@@ -83,16 +83,16 @@ To works correctly, this library needs some specific environment variables:
 * `BACKOFFICE_HEADER_KEY`
 * `MICROSERVICE_GATEWAY_SERVICE_NAME`
 
-If you create the service from the Platform, they are already defined. To change them you have to [do from DevOps console](https://docs.mia-platform.eu/development_suite/api-console/api-design/services#environment-variable-configuration).  
+When creating a new service  from Mia-Platform DevOps Console, they come already defined but you can always change or add them anytime as described [in the DevOps console documentation](https://docs.mia-platform.eu/development_suite/api-console/api-design/services#environment-variable-configuration).
```
Comment (fredmaggiowski) [inline README.md:86]:
```suggestion
When creating a new service from Mia-Platform DevOps Console, they come already defined but you can always change or add them anytime as described [in the DevOps console documentation](https://docs.mia-platform.eu/development_suite/api-console/api-design/services#environment-variable-configuration).  
```

File: docs/CustomService.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (fredmaggiowski) [reply docs/CustomService.md:18]: ok

File: docs/Decorators.md [hunk 13]
```diff
@@ -60,28 +58,24 @@ The first parameter of the handler function is [Request](https://www.fastify.io/
...
-    else {
-      const notifier=new Notifier()
-      // Try to send a notification
-      const response = await notifier.send({ text: `${who} says: ${mymsg}`})
-      // Adds to original response body the time of notification send
-      return request.changeOriginalResponse().setBody(
-          {...req.getOriginalRequestBody(),notifySendedAt:new Date() })
-    }
-) 
+    const notifier=new Notifier()
```
Comment (fredmaggiowski) [inline docs/Decorators.md:66]:
```suggestion
    const notifier = new Notifier()
```

File: docs/Decorators.md [hunk 14]
```diff
@@ -60,28 +58,24 @@ The first parameter of the handler function is [Request](https://www.fastify.io/
...
+    // Try to send a notification
+    const response = await notifier.send({ text: `${who} says: ${mymsg}`})
+    // Adds to original response body the time of notification send
+    return request.changeOriginalResponse().setBody(
+        { ...req.getOriginalRequestBody(),notifySendedAt:new Date() })
```
Comment (fredmaggiowski) [inline docs/Decorators.md:71]:
```suggestion
        { ...req.getOriginalRequestBody(),notifySendedAt:new Date() }
    )
```

File: README.md (hunk 11 above)
Comment (riccardozambito) [reply README.md:68]: @fredmaggiowski I have recovered the definition from Docs, give me feedback:

"The library exports a function which creates the infrastructure ready to accept the definition of routes and decorators. Optionally can take a schema of the required environment variables you can find the reference [here]. The function returned, `customService`, expects an async function to initialize and configure the `service`. "

Comment (fredmaggiowski) [reply README.md:68]: Sounds good! Let's use this definition!!

**Pull request merged**